File: scripts/monitoring_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import LaserScan
from geometry_msgs.msg import Twist
from assignment2_rt.msg import ObstacleInfo   # Import from YOUR package
from assignment2_rt.srv import SetThreshold   # Import from YOUR package
import math


class MonitoringNode(Node):
    def __init__(self):
        super().__init__('monitoring_node')

        self.safety_threshold = 1.0
        self.last_cmd_vel = Twist()

        self.create_subscription(LaserScan, '/scan', self.scan_callback, 10)
        self.create_subscription(Twist, '/cmd_vel', self.vel_callback, 10)

        self.pub_cmd = self.create_publisher(Twist, '/cmd_vel', 10)
        self.pub_info = self.create_publisher(ObstacleInfo, '/obstacle_info', 10)
        self.create_service(SetThreshold, 'set_safety_threshold', self.set_threshold_callback)

        self.get_logger().info("Monitoring Node Started.")

    def vel_callback(self, msg):
        self.last_cmd_vel = msg

    def set_threshold_callback(self, request, response):
        self.safety_threshold = request.new_threshold
        response.success = True
        return response

    def get_sector_name(self, index, total_len):
        one_third = total_len / 3
        if index < one_third:
            return "Right"
        elif index < 2 * one_third:
            return "Front"
        else:
            return "Left"

    def scan_callback(self, msg):
        valid_ranges = [r for r in msg.ranges if not math.isinf(r) and not math.isnan(r)]
        if not valid_ranges:
            return

        min_distance = min(valid_ranges)
        try:
            min_index = msg.ranges.index(min_distance)
            direction = self.get_sector_name(min_index, len(msg.ranges))
        except ValueError:
            direction = "Unknown"

        # Dashboard Display
        print(f"\033[2J\033[H", end="")
        print("=" * 40)
        print("       MONITORING DASHBOARD       ")
        print("=" * 40)
        print(
            f" COMMAND: Lin={
                self.last_cmd_vel.linear.x:.2f}, Ang={
                self.last_cmd_vel.angular.z:.2f}")
        print("-" * 40)
        print(f" SENSOR:  Min Dist={min_distance:.2f}m ({direction})")
        print(f" LIMIT:   Threshold={self.safety_threshold:.2f}m")
        print("-" * 40)

        if min_distance < self.safety_threshold:
            print(f" STATUS:  [!!! COLLISION RISK !!!]")
            stop_msg = Twist()
            if direction == "Front":
                stop_msg.linear.x = -1.0
            elif direction == "Left":
                stop_msg.angular.z = -1.0
            elif direction == "Right":
                stop_msg.angular.z = 1.0
            else:
                stop_msg.linear.x = 1.0
            self.pub_cmd.publish(stop_msg)
        else:
            print(f" STATUS:  [ OK ] Safe")
        print("=" * 40)

        info_msg = ObstacleInfo()
        info_msg.distance = float(min_distance)
        info_msg.direction = direction
        info_msg.threshold = float(self.safety_threshold)
        self.pub_info.publish(info_msg)


def main(args=None):
    rclpy.init(args=args)
    node = MonitoringNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/monitoring_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "assignment2_rt/msg/obstacle_info.hpp"
#include "assignment2_rt/srv/set_threshold.hpp"

#include <algorithm>
#include <cmath>
#include <limits>

class MonitoringNode : public rclcpp::Node
{
public:
    MonitoringNode() : Node("monitoring_node"), safety_threshold_(1.0), reversing_(false)
    {
        // 1. Subscribers & Publishers
        sub_scan_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
            "/scan", 10, std::bind(&MonitoringNode::scan_callback, this, std::placeholders::_1));
        
        pub_cmd_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);
        pub_info_ = this->create_publisher<assignment2_rt::msg::ObstacleInfo>("/obstacle_info", 10);

        // 2. Service
        srv_threshold_ = this->create_service<assignment2_rt::srv::SetThreshold>(
            "set_safety_threshold",
            std::bind(&MonitoringNode::set_threshold_callback, this, std::placeholders::_1, std::placeholders::_2));

        RCLCPP_INFO(this->get_logger(), "Monitoring Node (C++) Started. Threshold: 1.0m");
    }

private:
    void set_threshold_callback(const std::shared_ptr<assignment2_rt::srv::SetThreshold::Request> request,
                                std::shared_ptr<assignment2_rt::srv::SetThreshold::Response> response)
    {
        safety_threshold_ = request->new_threshold;
        response->success = true;
        RCLCPP_INFO(this->get_logger(), "Threshold updated to: %.2fm", safety_threshold_);
    }

    std::string get_sector_name(size_t index, size_t total_len)
    {
        size_t one_third = total_len / 3;
        if (index < one_third) return "Right";
        else if (index < 2 * one_third) return "Front";
        else return "Left";
    }

    void scan_callback(const sensor_msgs::msg::LaserScan::SharedPtr msg)
    {
        // Find minimum distance (ignoring inf/nan)
        float min_distance = std::numeric_limits<float>::infinity();
        size_t min_index = 0;
        bool found_valid = false;

        for (size_t i = 0; i < msg->ranges.size(); ++i) {
            float r = msg->ranges[i];
            if (!std::isinf(r) && !std::isnan(r)) {
                if (r < min_distance) {
                    min_distance = r;
                    min_index = i;
                    found_valid = true;
                }
            }
        }

        if (!found_valid) return;

        std::string direction = get_sector_name(min_index, msg->ranges.size());

        // Publish Info
        auto info_msg = assignment2_rt::msg::ObstacleInfo();
        info_msg.distance = min_distance;
        info_msg.direction = direction;
        info_msg.threshold = safety_threshold_;
        pub_info_->publish(info_msg);

        // Safety Logic
        auto twist_msg = geometry_msgs::msg::Twist();

        if (min_distance < safety_threshold_) {
            RCLCPP_WARN(this->get_logger(), "OBSTACLE DETECTED in %s! Moving Back...", direction.c_str());
            reversing_ = true;
            twist_msg.linear.x = -0.5;
            twist_msg.angular.z = 0.0;
            pub_cmd_->publish(twist_msg);
        }
        else if (reversing_) {
            // Obstacle cleared, stop backing up
            RCLCPP_INFO(this->get_logger(), "Safe distance restored. Stopping robot.");
            reversing_ = false;
            twist_msg.linear.x = 0.0;
            twist_msg.angular.z = 0.0;
            pub_cmd_->publish(twist_msg);
        }
    }

    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr sub_scan_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr pub_cmd_;
    rclcpp::Publisher<assignment2_rt::msg::ObstacleInfo>::SharedPtr pub_info_;
    rclcpp::Service<assignment2_rt::srv::SetThreshold>::SharedPtr srv_threshold_;

    float safety_threshold_;
    bool reversing_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MonitoringNode>());
    rclcpp::shutdown();
    return 0;
}
